rfcs/20221124-monorepo.md

```diff
@@ -248,7 +280,29 @@ To clean up the workflows we will define three new actions. The code for all thr
 
 #### Update workflows
 
-With this done, the development on the API and frontend can continue inside their subdirectories. The development of both parts will be independent. At least until we reach [long-term consolidation](#step-5-long-term-consolidation).
+Workflows with major changes:
+
+- `ci_cd.yml` from the API will absorb `ci.yml` from the frontend
+- `lint.yml` will be deleted
+
+Updates:
+
+- `migration_safety_warning.yml`
+- `generate_pot.yml`
+
+With this done, the development on the API and frontend can continue inside their subdirectories. The development of both parts will be independent, at least until we reach [long-term consolidation](#step-5-long-term-consolidation).
+
+#### Deployment
+
+##### Staging
+
+The soon-to-be-ECS-based API and ECS-based frontend will continue to deploy to staging via the CI + CD pipeline, with deployment starting as soon as all CI checks have passed. They will use similar code as the frontend auto-deploy for staging used currently.
+
+These will be separate jobs with specific [path-based filters](https://github.com/dorny/paths-filter).
```

Comment:
What is the benefit of using this `paths-filter` action rather than the built-in GitHub workflow `paths` settings for workflow triggers? Similar to my other question: https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#using-filters-to-target-specific-paths-for-pull-request-or-push-events

Reply:
This actions allows us to filter jobs instead of applying the filter to the entire workflow. This allows us to have a complete workflow where the job to deploy images to GHCR can run if all the CI test jobs have passed.

If the deploy is in a separate workflow, I don't think it's possible to run a workflow if a different workflow has passed.

Reply:
It is possible to have cross-workflow dependencies, actually: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run

I only have two concerns with using the actions:

1. Extremely repetitive and annoying code to maintain in the workflows (need to guard each frontend/api job with the same repeated step at the top of the job). Note that we could use YAML anchors to fix this _slightly_ (still more repetition here compared to `paths`) but [`actionlint` does not currently support them](https://github.com/rhysd/actionlint/issues/133). We'd either need to contribute an upstream PR or stop using `actionlint`, the latter of which I think is unwise given how helpful it is at catching syntax errors locally.
2. It still actually uses a runner and worker time to complete whereas needs bypasses it entirely, therefore not eating up any of our quota. Considering the sheer number of jobs we're going to end up having in a single repository by the end of this migration, cutting down on the number of jobs that are run is prudent in this respect. This is one reason why we added all the concurrency configuration to our workflows, for example.

If you think either of these are not a concern, then that is fine. The first should be documented as an expectation and accepted trade off. I think the second would bite us later on, but I can accept that I may be being _too_ cautious.

One more thing that crossed my mind as I was writing this is whether our concurrency settings will need to change for jobs. My hunch is _no_, but it's worth looking into to confirm as they can be quite nasty to think through.

Reply:
I agree that `actionlint` has a lot of value so losing it would not be advisable. Let me update some of the code in `monoverse` to use `workflow_run`. Hopefully that'll allow us to keep the jobs distributed.

Reply:
`workflow_run` workflows depend on the ["OR" combination](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run) of multiple workflows. This is different from `needs` which depends the ["AND" combination](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idneeds) of multiple jobs.

The annoying to maintain code might be true, but it can be overcome by abstracting the `paths-filter` job into a separate action with some `inputs` (similar to the `setup-env` action), thus limiting the annoyance to a single file.

Point 2 could be an issue as it will spin up a job and cancel it if the criterion fails, an alternative to that could be to have a job that only matches the condition of the paths. Let's say jobs called "is_api", "is_frontend" etc. The other jobs can `needs` on them so that they're skipped if their code paths are untouched. I think it will be a bit of a hassle to manage the pass/fail/skip cascades though so it's something that needs a bit of a dry-run first.